File: app/src/main/java/bkm/com/newsapp/utilities/InjectorUtils.java
package bkm.com.newsapp.utilities;

import android.content.Context;

import bkm.com.newsapp.data.NewsRepository;
import bkm.com.newsapp.data.database.NewsDatabase;
import bkm.com.newsapp.data.network.NewsNetworkDataSource;
import bkm.com.newsapp.ui.list.MainViewModelFactory;

public class InjectorUtils {

    public static NewsRepository provideRepository(Context context) {
        NewsDatabase database = NewsDatabase.getInstance(context.getApplicationContext());
        AppExecutors executors = AppExecutors.getInstance();
        NewsNetworkDataSource networkDataSource = NewsNetworkDataSource.getInstance(context.getApplicationContext(), executors);
        return NewsRepository.getInstance(database.newsDao(), networkDataSource, executors);
    }

    public static MainViewModelFactory provideMainActivityViewModelFactory(Context context) {
        NewsRepository repository = provideRepository(context.getApplicationContext());
        return new MainViewModelFactory(repository);
    }

    public static NewsNetworkDataSource provideNetworkDataSource(Context context) {
        provideRepository(context.getApplicationContext());
        AppExecutors executors = AppExecutors.getInstance();
        return NewsNetworkDataSource.getInstance(context.getApplicationContext(), executors);
    }
}

File: app/src/main/java/bkm/com/newsapp/utilities/InjectorUtils.kt
package bkm.com.newsapp.utilities

import android.content.Context
import bkm.com.newsapp.data.NewsRepository
import bkm.com.newsapp.data.database.NewsDatabase
import bkm.com.newsapp.data.network.NewsNetworkDataSource
import bkm.com.newsapp.ui.list.MainViewModelFactory

class InjectorUtils {
    companion object {
        fun provideRepository(context: Context): NewsRepository {
            var database = NewsDatabase.getInstance(context.applicationContext)
            var executors = AppExecutors.getInstance()
            var networkDataSource = NewsNetworkDataSource.getInstance(context, executors)
            return NewsRepository.getInstance(database.newsDao(), networkDataSource, executors)
        }

        fun provideMainActivityViewModelFactory(context: Context): MainViewModelFactory {
            var repository = provideRepository(context.applicationContext)
            return MainViewModelFactory(repository)
        }

        fun provideNetworkDataSource(context: Context): NewsNetworkDataSource? {
            provideRepository(context.applicationContext)
            var executors = AppExecutors.getInstance()
            return NewsNetworkDataSource.getInstance(context.applicationContext, executors)
        }
    }
}